app/code/Magento/CatalogUrlRewrite/Model/ProductScopeRewriteGenerator.php

```diff
@@ -276,4 +283,83 @@ private function isCategoryRewritesEnabled()
     {
         return (bool)$this->config->getValue('catalog/seo/generate_category_product_rewrites');
     }
+
+    /**
+     * Return product changed categories store root category ids
+     *
+     * @param Product $product
+     * @return int[]
+     */
+    private function getChangedCategoriesStoreIds(Product $product): array
+    {
+        $oldCategoryIds = $product->getOrigData('category_ids');
+        $actualCategoryIds = $product->getCategoryIds();
+        $result = [];
+        if ($oldCategoryIds === null) {
+            return $result;
+        }
+        if (array_diff($actualCategoryIds, $oldCategoryIds)) {
+            $result = $this->getAddedCategoriesStoreIds($product);
+        }
+        if (array_diff($oldCategoryIds, $actualCategoryIds)) {
+            $result = array_merge($result, $this->getRemovedCategoriesStoreIds($product));
+        }
+
+        return array_map('intval', array_unique($result));
+    }
+
+    /**
+     * Return added product categories store root category ids
+     *
+     * @param Product $product
+     * @return array
+     */
+    private function getAddedCategoriesStoreIds(Product $product): array
+    {
+        $newCategories = array_filter(
+            $product->getCategoryCollection()->getItems(),
+            function (CategoryInterface $category) use ($product) {
+                return !in_array($category->getId(), $product->getOrigData('category_ids'));
+            }
+        );
+
+        return $this->getCategoriesStoreIds($newCategories);
+    }
+
+    /**
+     * Return removed product categories store root category ids
+     *
+     * @param Product $product
+     * @return array
+     */
+    private function getRemovedCategoriesStoreIds(Product $product): array
+    {
+        $categoriesCollection = clone $product->getCategoryCollection();
+        $categoriesCollection->clear()->getSelect()->reset(Select::WHERE);
+        $categoriesCollection->addIdFilter(
+            array_diff($product->getOrigData('category_ids'), $product->getCategoryIds())
+        );
+
+        return $this->getCategoriesStoreIds($categoriesCollection->getItems());
+    }
+
+    /**
+     * Return store root category id parsed from path
+     *
+     * @param Category[] $categories
+     * @return array
+     */
+    private function getCategoriesStoreIds(array $categories): array
+    {
+        $storeIds = array_reduce(
+            $categories,
+            function (array $result, Category $category) {
+                $result[] = $category->getStoreIds();
+
+                return $result;
+            },
+            []
+        );
+        return array_merge([], ...$storeIds);
```

Comment:
Probably could be simplified to
```suggestion
        $result = [[]];
        foreach ($categories as $category) {
            $result[] = $category->getStoreIds();
        }

        $result = array_merge(...$result);

        return array_unique($result);
```